sw/device/tests/rstmgr_cpu_info_test.c

```diff
@@ -24,176 +23,253 @@ OTTF_DEFINE_TEST_CONFIG();
  * RSTMGR CPU INFO TEST
  *  This test creates a double fault by accessing a register with
  *  a non-existing address.
- *  After the double fault, the dut gets reset by the watch dog bite,
+ *  After a single fault and then a double fault,
+ *  the dut gets reset by the watch dog bite,
  *  and the test collects / checks the cpu_info from the rstmgr.
  */
 
-// non existing address
-#define kIllegalAddr1 0x4041FFF0u
-#define kIllegalAddr2 0x40003618u
-#define kSkipComp 0x12345678u
+// Unmapped Addresses
+static const uint8_t *kIllegalAddr0 = (uint8_t *)0xA0A0DEAFu;
+static const uint8_t *kIllegalAddr1 = (uint8_t *)0xA041FFF0u;
+static const uint8_t *kIllegalAddr2 = (uint8_t *)0xA0003618u;
 #define kCpuDumpSize 8
 
+// Declaring the labels used to calculate the expected current and next pc
+// after a double fault.
+extern const uint32_t _ottf_interrupt_vector, handler_exception;
+
+// The labels to points in the code of which the memory address is needed.
+extern const char kSingleFaultAddrLower[], kSingleFaultAddrUpper[],
+    kSingleFaultAddrCurrentPc[], kSingleFaultAddrNextPc[],
+    kDoubleFaultFirstAddrLower[], kDoubleFaultFirstAddrUpper[],
+    kDoubleFaultSecondAddrLower[], kDoubleFaultSecondAddrUpper[];
+
+// A handle to the reset manager.
+static dif_rstmgr_t rstmgr;
+
 /**
- * Cpu dump struct index
+ * This variable is used to ensure loads from an address aren't optimised out.
  */
-enum {
-  kCpuDumpIdxCurrentExceptionAddr = 0,
-  kCpuDumpIdxCurrentExceptionPc = 1,
-  kCpuDumpIdxCurrentLastDataAddr = 2,
-  kCpuDumpIdxCurrentNextPc = 3,
-  kCpuDumpIdxCurrentPc = 4,
-  kCpuDumpIdxPreviousExceptionAddr = 5,
-  kCpuDumpIdxPreviousExceptionPc = 6,
-  kCpuDumpIdxPreviousValid = 7,
-};
+volatile uint8_t addr_val;
 
 /**
- * Reserve expected cpu dump area in flash
+ * When true, the exception handler will trigger another fault,
+ * causing a double fault,
+ * otherwise it triggers a software reset.
  */
-OT_SECTION(".non_volatile_scratch")
-dif_rstmgr_cpu_info_dump_segment_t exp_dump[kCpuDumpSize];
+static bool double_fault;
```

Comment:
This is shared between the handler and the main program, thus should be marked as volatile.